File: email_validator.py
import re
import sys

class Email_Validator:
    def __init__(self):
        self.name_pattern = re.compile(r"^[a-zA-Z0-9!#$%&'*+\-/=?^_`{|}~.]+$")
        self.domain_pattern = re.compile(r"^[a-zA-Z0-9.-]+$")

    def validate(self, email: str) -> bool:
        if email == "" or len(email) > 254:
            return False, "Failed: Length exceeds 254 characters or is empty."

        if email.count('@') != 1: 
            return False
        name_part, domain_part = email.rsplit('@', 1)

        if name_part == "" or len(name_part) > 64:
            return False

        if name_part.startswith('.') or name_part.endswith('.'):
            return False

        if domain_part.startswith('.') or domain_part.endswith('.'):
            return False

        if ".." in name_part:
            return False

        if ".." in domain_part:
            return False

        is_quoted = (
            name_part.startswith('"') and 
            name_part.endswith('"')
        )
        name_test = name_part[1:-1] if is_quoted else name_part

        if not is_quoted and not self.name_pattern.fullmatch(name_test):
            return False

        if not self.domain_pattern.fullmatch(domain_part):
            return False

        if '.' not in domain_part:
            return False

        return True, "Valid email format."


if __name__ == "__main__":
    # Check if an email was passed as an argument
    if len(sys.argv) < 2:
        print("[-] Usage: python3 email_validator.py <email_address>")
        sys.exit(1)
        
    validator = Email_Validator()
    
    # Grab the exact string passed in the terminal
    email = sys.argv[1]
    
    is_valid = validator.validate(email)
    status = "\033[92m[Valid]\033[0m" if is_valid else "\033[91m[Invalid]\033[0m"
    print(f"{status} {email}")

File: email_validator.cpp
#include <iostream>
#include <string>
#include <regex>
#include <algorithm>

class Email_validator {
private:
    std::regex name_pattern;
    std::regex domain_pattern;

public:
    Email_validator() {
        name_pattern = std::regex("^[a-zA-Z0-9!#$'*+\\-/=?^_`{|}~.]+$");
        domain_pattern = std::regex("^[a-zA-Z0-9.-]+$");
    }

    bool validate(const std::string& email) {
        if (email.empty() || email.length() > 254) return false;
        if (std::count(email.begin(), email.end(), '@') != 1) return false;
        
        size_t spliting_the_whole_mail = email.find('@');

        std::string name_part = email.substr(0, spliting_the_whole_mail);
        std::string domain_part = email.substr(spliting_the_whole_mail + 1);

        if (name_part.empty() || name_part.length() > 64) return false;
        if (name_part.front() == '.' || name_part.back() == '.') return false;
        if (domain_part.front() == '.' || domain_part.back() == '.') return false;
        if (name_part.find("..") != std::string::npos) return false;
        if (domain_part.find("..") != std::string::npos) return false;

        bool is_quoted = false;
        if (name_part.length() >= 2 && name_part.front() == '"' && name_part.back() == '"') {
            is_quoted = true;
        }
        
        std::string name_test = is_quoted ? name_part.substr(1, name_part.length() - 2) : name_part;

        if (!is_quoted && !std::regex_match(name_test, name_pattern)) return false;
        if (!std::regex_match(domain_part, domain_pattern)) return false;
        if (domain_part.find('.') == std::string::npos) return false;

        return true;
    }
};

int main(int argc, char* argv[]) {
    // Check if an email was passed as an argument
    if (argc < 2) {
        std::cerr << "[-] Usage: ./validator <email_address>\n";
        return 1;
    }

    Email_validator validator;
    
    // Grab the exact string passed in the terminal
    std::string email = argv[1];
    
    bool isValid = validator.validate(email);
    
    if (isValid) {
        std::cout << "\033[92m[Valid]\033[0m " << email << "\n";
    } else {
        std::cout << "\033[91m[Invalid]\033[0m " << email << "\n";
    }

    return 0;
}